### PR TITLE
Support instance "providers"

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Replaces previous flawed approach from #10.

### CODE REVIEW

File: tests/pipeline_test.py [hunk 1]
```diff
@@ -388,3 +388,98 @@ def provide_C(x: A[T1], y: B[T2]) -> C[T1, T2]:
...
+    assert pl.compute(float) == 2.0
+    assert pl.compute(Result) == 1.5
+
+
+def test_provider_NewType_instance() -> None:
+    A = NewType('A', int)
+    pl = sl.Pipeline([])
+    pl[A] = A(3)
+    assert pl.compute(A) == 3
+
```
Comment (YooSunYoung) [inline tests/pipeline_test.py:412]: Maybe this kind of test case can be useful...?

```suggestion

def test_provider_Newtype_instance_same_supertype() -> None:
    pl = sl.Pipeline()
    Str = NewType('Str', str)
    IntStr = NewType('IntStr', str)
    pl[IntStr] = Str('0')
    assert pl.compute(IntStr) == '0'

```

Comment (SimonHeybrock) [reply tests/pipeline_test.py:412]: I suppose this is just a "limitation" of how `NewType` works. I think the type checker should be able to catch this?

Comment (SimonHeybrock) [reply tests/pipeline_test.py:412]: `mypy` does not, apparently. :(

Comment (YooSunYoung) [reply tests/pipeline_test.py:412]: This one is not caught by `mypy` or `pyre` either.
`TypeVar` seems quite loose if it's not used in `Generic` ...
If your intention is not allowing this at all, maybe adding it to the document should be sufficient...?

```
from typing import NewType, Dict, Type, TypeVar, Any
A = NewType('A', str)
B = NewType('B', str)

T = TypeVar("T")

class C:
    def __init__(self) -> None:
        self.type_dict: Dict[Any, Any] = dict()
    
    def __setitem__(self, tp: Type[T], _object: T) -> None:
        self.type_dict[tp] = _object

c = C()
c[A] = B('str')
c[int] = '0'
```

Comment (SimonHeybrock) [reply tests/pipeline_test.py:412]: > If your intention is not allowing this at all, maybe adding it to the document should be sufficient...?

Not allowing what?

Comment (YooSunYoung) [reply tests/pipeline_test.py:412]: To do this, setting an item of type `Str` as a provider of `IntStr` in this example.
```
 pl = sl.Pipeline()
Str = NewType('Str', str)
IntStr = NewType('IntStr', str)
pl[IntStr] = Str('0')
assert pl.compute(IntStr) == '0'
```

**Review by YooSunYoung — APPROVED** [review]

File: tests/pipeline_test.py (hunk 1 above)
Comment (SimonHeybrock) [reply tests/pipeline_test.py:412]: Well, I don't see how I could prevent that, except by disallowing `NewType` as key?

**Pull request merged**